### PR TITLE
New job, node and agent CLIs

### PR DESCRIPTION
**Pull request by wdbaruni**
This PR introduces new CLIs to interact with the [new set of APIs](https://github.com/bacalhau-project/bacalhau/pull/2791) that allows submitting new job types.

## Notes
1. Existing CLIs, such as `bacalhau describe` `bacalhau docker run` and `bacalhau create` are still working as before, calling existing APIs and using existing job models.
2. New CLIs are grouped based on the resource they are operating on, such as `bacalhau job ...` and `bacalhau node ...`
3. `bacalhau job describe [id]` only prints the job spec and its current state, and unlink existing `bacalhau describe`, it doesn't show executions or history events. To get executions you need to call `bacalhau job executions [id]` and to get history events you need to call `bacalhau job history [id]`. This needs to be revisited and allow `bacalhau job describe [id]` to show a summary of the job's spec, its executions and possibly evaluations.
4. There is no equivalent to `bacalhau get [id]` to download results from IPFS. This is an area that requires more thought and whether downloading results is part of our CLI or not
5. Publisher used to be a required field where we use IPFS by default if none is specified. This PR makes publishers required only if the user defined outputs in their job spec. Otherwise no results will be published and users can still use `bacalhau logs [id]` to query logs
6. Similar to execution IDs that are prefixed with `e-`, job UUIDs are now prefixed with `j-`, evaluations are prefixed with `v-`
7. This PR also introduces [job label filtering](https://github.com/bacalhau-project/bacalhau/issues/1742)

### Jobs
```
# Submit job from file
bacalhau job run testdata/jobs/docker.yaml

# Submit job from stdin
cat testdata/jobs/docker.yaml | bacalhau job run

# Re-submit a previous job
bacalhau job describe j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79 | bacalhau job run

# List jobs
bacalhau job list

# List jobs and filter by labels to 
bacalhau job list --labels "env=production,\!canary"

# List jobs in yaml
bacalhau job list --output yaml

# List jobs in pretty json
bacalhau job list --output json --pretty

# Describe job, which includes the submitted spec, state and revision
bacalhau job describe j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79

# List job executions 
bacalhau job executions j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79

# List job executions in yaml
bacalhau job executions j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79 --output yaml

# List job history
bacalhau job history j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79

# List execution level history events
bacalhau job history j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79 --event-type execution

# List history for a specific execution
bacalhau job history j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79 --execution-id e-19af2b35

# List history for a specific node id
bacalhau job history j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79 --node-id Qmc3mb8b

# Query logs
bacalhau job logs j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79

# Stop (aka Cancel) job
bacalhau job stop j-7720bb8f-6dfc-49b2-8c9a-5359cbe7ce79
```

### Agent
Set of commands to query the bacalhau agent (node) the client is connected to using `--api-host` and `--api-port`. Not the client binary
```
# Version of bacalhau deployed on the node
bacalhau agent version

# Print node info
bacalhau agent node

# Node liveness and health info
bacalhau agent alive
```

### Node
Set of commands to query info about other nodes in the network 

```
# List all nodes
bacalhau node list

# Filter nodes by labels
bacalhau node list --labels "env=devstack,region=us-east-1"

# Order by available CPU and output in pretty json
bacalhau node list --order-by available_cpu --output json --pretty

# Describe node by ID
 bacalhau node describe QmbkBjSraDGpdG3a4t9hg9k5iFFc6xiJn3Tu1dX17GyMgQ 
```

### CODE REVIEW

**Review by rossjones — APPROVED** [review]
Looking good, I think I like the new approach to namespacing the commands.  A few comments where the CLI runtime docs don't have the correct object prefix on IDs, and a couple of others that are hopefully useful.

Something I discovered from manually running all of the commands in the PR description, is that although it often shows the short ID, it doesn't accept it.  Perhaps a ticket to make sure we go back and enable it?  

For example 

``` shell
$ ./bin/darwin_arm64/bacalhau node list
ID                Type           Lablews
QmYxziHS  Compute    env=devstack                                        6.4 /   19.2 GB /   248.7 GB /   0 /

$ ./bin/darwin_arm64/bacalhau node describe QmYxziHS
could not get node QmYxziHS: Unexpected response code: 400 ({
  "error": "code=400, message=failed to parse peer ID: length greater than remaining number of bytes in buffer",
  "message": "failed to parse peer ID: length greater than remaining number of bytes in buffer"
})

$ ./bin/darwin_arm64/bacalhau node describe QmYxziHS2F1N32xpur1cCvRb4XKv3wHc7cUzTVK1f81YZY
peerinfo:
  id: QmYxziHS2F1N32xpur1cCvRb4XKv3wHc7cUzTVK1f81YZY
  addrs: []
nodetype: Compute
labels:
  env: devstack
  id: QmYxziHS2F1N32xpur1cCvRb4XKv3wHc7cUzTVK1f81YZY
  name: node-3
.....
```

File: cmd/cli/job/logs.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+var (
+	logsShortDesc = templates.LongDesc(i18n.T(`
+		Follow logs from a currently executing job
+`))
+
+	logsExample = templates.Examples(i18n.T(`
+		# Follow logs for a previously submitted job
+		bacalhau logs 51225160-807e-48b8-88c9-28311c7899e1
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/logs.go]: fixed in [c014328](https://github.com/bacalhau-project/bacalhau/pull/2802/commits/c0143285d7e77c68c610acc8457652c82a5bbedb)

File: cmd/cli/job/logs.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+		# Follow output with a short ID 
+		bacalhau logs ebd9bf2f
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/logs.go]: yeah the `idgen.ShortID` implementation preserves the prefix if it exists, but not the old `models.ShortID` implementation. I've fixed this in [c014328](https://github.com/bacalhau-project/bacalhau/pull/2802/commits/c0143285d7e77c68c610acc8457652c82a5bbedb) and stopped relying on the ID length to know if it is short or not

File: cmd/cli/job/describe.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	describeLong = templates.LongDesc(i18n.T(`
+		Full description of a job, in yaml format. 
+		Use 'bacalhau job list' to get a list of jobs.
+`))
+	describeExample = templates.Examples(i18n.T(`
+		# Describe a job with the full ID
+		bacalhau job describe e3f8c209-d683-4a41-b840-f09b88d087b9
+
+		# Describe a job with the a shortened ID
+		bacalhau job describe 47805f5c
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/describe.go]: yeah you are right. fixed in [c014328](https://github.com/bacalhau-project/bacalhau/pull/2802/commits/c0143285d7e77c68c610acc8457652c82a5bbedb)

File: cmd/cli/job/stop.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+var (
+	stopLong = templates.LongDesc(i18n.T(`
+		Stop a previously submitted job.
+`))
+
+	//nolint:lll // Documentation
+	stopExample = templates.Examples(i18n.T(`
+		# Stop a previously submitted job
+		bacalhau job stop 51225160-807e-48b8-88c9-28311c7899e1
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/stop.go]: fixed in [c014328](https://github.com/bacalhau-project/bacalhau/pull/2802/commits/c0143285d7e77c68c610acc8457652c82a5bbedb)

File: pkg/models/task.go [hunk 5]
```diff
@@ -88,7 +92,23 @@ func (t *Task) Copy() *Task {
 	return nt
 }
 
+// Validate is used to check a job for reasonable configuration
 func (t *Task) Validate() error {
+	mErr := new(multierror.Error)
+	mErr = multierror.Append(mErr, t.ValidateSubmission())
+
+	if err := t.Timeouts.Validate(); err != nil {
+		mErr.Errors = append(mErr.Errors, fmt.Errorf("task timeouts validation failed: %v", err))
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply pkg/models/task.go]: yeah both works, but `multierror.Append` is better as it allows appending and flattening multierrors if they are appended to each other. Switched to using `multierror.Append` in [c014328](https://github.com/bacalhau-project/bacalhau/pull/2802/commits/c0143285d7e77c68c610acc8457652c82a5bbedb)

File: cmd/cli/job/list.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		Short:   listShort,
+		Long:    listLong,
+		Example: listExample,
+		PreRun:  util.ApplyPorcelainLogLevel,
+		Args:    cobra.NoArgs,
+		Run:     o.run,
+	}
+
+	listCmd.Flags().StringVar(&o.Labels, "labels", o.Labels,
+		"Filter nodes by labels. See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ for more information.")
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/list.go:75]: Done https://expanso.atlassian.net/browse/ENG-103

File: cmd/cli/node/list.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		Short:  "List info of network nodes. ",
+		PreRun: util.ApplyPorcelainLogLevel,
+		Args:   cobra.NoArgs,
+		Run:    o.run,
+	}
+	nodeCmd.Flags().StringSliceVar(&o.ColumnGroups, "show", o.ColumnGroups,
+		fmt.Sprintf("What column groups to show. Zero or more of: %q", maps.Keys(toggleColumns)))
+
+	nodeCmd.Flags().StringVar(&o.Labels, "labels", o.Labels,
+		"Filter nodes by labels. See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ for more information.")
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/node/list.go:48]: Done https://expanso.atlassian.net/browse/ENG-103

**Pull request merged**